### PR TITLE
added book model and repository

### PR DESCRIPTION
**Pull request by Litvynets**

### CODE REVIEW

File: src/main/java/mate/academy/bookstore/model/Book.java [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+import jakarta.persistence.Entity;
+import jakarta.persistence.GeneratedValue;
+import jakarta.persistence.GenerationType;
+import jakarta.persistence.Id;
+import jakarta.persistence.Table;
+import java.math.BigDecimal;
+import lombok.Data;
+
+@Entity
+@Data
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/bookstore/model/Book.java:13]:
```suggestion
@Getter
@Setter
```

File: src/main/java/mate/academy/bookstore/repository/BookRepository.java [hunk 2]
```diff
@@ -0,0 +1,13 @@
...
+import java.util.List;
+import mate.academy.bookstore.model.Book;
+import org.springframework.data.jpa.repository.JpaRepository;
+import org.springframework.stereotype.Repository;
+
+@Repository
+public interface BookRepository extends JpaRepository<Book, Long> {
+    Book save(Book book);
+
+    List<Book> findAllBy();
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/bookstore/repository/BookRepository.java:12]:
```suggestion
    List<Book> findAll();
```

File: src/main/java/mate/academy/bookstore/service/BookService.java [hunk 3]
```diff
@@ -0,0 +1,12 @@
+package mate.academy.bookstore.service;
+
+import java.util.List;
+import mate.academy.bookstore.model.Book;
+import org.springframework.stereotype.Service;
+
+@Service
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/bookstore/service/BookService.java:7]:
```suggestion
```

File: src/main/java/mate/academy/bookstore/repository/BookRepository.java [hunk 4]
```diff
@@ -0,0 +1,13 @@
+package mate.academy.bookstore.repository;
+
+import java.util.List;
+import mate.academy.bookstore.model.Book;
+import org.springframework.data.jpa.repository.JpaRepository;
+import org.springframework.stereotype.Repository;
+
+@Repository
+public interface BookRepository extends JpaRepository<Book, Long> {
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/bookstore/repository/BookRepository.java:9]:
```suggestion
public interface BookRepository {
```

File: src/main/java/mate/academy/bookstore/service/impl/BookServiceImpl.java [hunk 5]
```diff
@@ -0,0 +1,24 @@
...
+import org.springframework.stereotype.Service;
+
+@Service
+public class BookServiceImpl implements BookService {
+    @Autowired
+    private BookRepository bookRepository;
+
+    @Override
+    public Book save(Book book) {
+        return bookRepository.save(book);
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/bookstore/service/impl/BookServiceImpl.java:17]: You need to implement methods save and findAll in BookRepositoryImpl class
Please, don't use JpaRepository

**Review by Elena-Bruyako — CHANGES_REQUESTED** [review]
See comments

File: src/main/java/mate/academy/bookstore/model/Book.java [hunk 6]
```diff
@@ -0,0 +1,33 @@
...
+import jakarta.persistence.GenerationType;
+import jakarta.persistence.Id;
+import jakarta.persistence.Table;
+import java.math.BigDecimal;
+import lombok.Getter;
+import lombok.RequiredArgsConstructor;
+import lombok.Setter;
+
+@Entity
+@RequiredArgsConstructor
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/bookstore/model/Book.java:15]:
```suggestion
```

File: src/main/java/mate/academy/bookstore/repository/impl/BookRepositoryImpl.java [hunk 7]
```diff
@@ -0,0 +1,68 @@
...
+import mate.academy.bookstore.model.Book;
+import mate.academy.bookstore.repository.BookRepository;
+import org.hibernate.Session;
+import org.hibernate.SessionFactory;
+import org.hibernate.Transaction;
+import org.hibernate.query.Query;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.stereotype.Repository;
+
+@Repository
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/bookstore/repository/impl/BookRepositoryImpl.java:16]:
```suggestion
@Repository
@RequiredArgsConstructor
```

File: src/main/java/mate/academy/bookstore/repository/impl/BookRepositoryImpl.java [hunk 8]
```diff
@@ -0,0 +1,68 @@
...
+public class BookRepositoryImpl implements BookRepository {
+    private final SessionFactory sessionFactory;
+
+    @Autowired
+    public BookRepositoryImpl(EntityManagerFactory entityManagerFactory) {
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/bookstore/repository/impl/BookRepositoryImpl.java:21]: remove

File: src/main/java/mate/academy/bookstore/repository/impl/BookRepositoryImpl.java [hunk 9]
```diff
@@ -0,0 +1,68 @@
...
+        try {
+            session = sessionFactory.openSession();
+            transaction = session.beginTransaction();
+            session.persist(book);
+            transaction.commit();
+        } catch (Exception e) {
+            if (transaction != null) {
+                transaction.rollback();
+            }
+            throw new RuntimeException("Can't save book: " + book + " to the DB", e);
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/bookstore/repository/impl/BookRepositoryImpl.java:38]:
```suggestion
            throw new DataProcessingException("Can't save book: " + book + " to the DB", e);
```

same for other places

File: src/main/java/mate/academy/bookstore/service/impl/BookServiceImpl.java [hunk 10]
```diff
@@ -0,0 +1,40 @@
...
+    @Override
+    public BookDto save(CreateBookRequestDto book) {
+        Book savedBook = bookRepository.save(bookMapper.toModel(book));
+        return bookMapper.toDto(savedBook);
+    }
+
+    @Override
+    public List<BookDto> findAll() {
+        List<Book> allBook = bookRepository.findAll();
+        return allBook.stream()
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/bookstore/service/impl/BookServiceImpl.java:29]:
```suggestion
        return bookRepository.findAll().stream()
```

File: src/main/java/mate/academy/bookstore/service/impl/BookServiceImpl.java [hunk 11]
```diff
@@ -0,0 +1,40 @@
...
+
+@Service
+@AllArgsConstructor
+public class BookServiceImpl implements BookService {
+    private BookMapper bookMapper;
+    private BookRepository bookRepository;
+
+    @Override
+    public BookDto save(CreateBookRequestDto book) {
+        Book savedBook = bookRepository.save(bookMapper.toModel(book));
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/bookstore/service/impl/BookServiceImpl.java:22]:
```suggestion
        Book book = bookRepository.save(bookMapper.toModel(book));
```

**Review by Elena-Bruyako — CHANGES_REQUESTED** [review]
see comments

File: src/main/java/mate/academy/bookstore/controller/BookController.java [hunk 12]
```diff
@@ -0,0 +1,35 @@
...
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.web.bind.annotation.GetMapping;
+import org.springframework.web.bind.annotation.PathVariable;
+import org.springframework.web.bind.annotation.PostMapping;
+import org.springframework.web.bind.annotation.RequestBody;
+import org.springframework.web.bind.annotation.RequestMapping;
+import org.springframework.web.bind.annotation.RestController;
+
+@RestController
+@RequestMapping("/api/books")
```
Comment (okuzan) [inline src/main/java/mate/academy/bookstore/controller/BookController.java:16]: move project-wide prefix `/api` to `server.servlet.context-path`

**Review by okuzan — CHANGES_REQUESTED** [review]

**Review by Elena-Bruyako — CHANGES_REQUESTED** [review]
Looks like this PR is not related to Spring Boot Data JPA HW

File: src/main/java/mate/academy/bookstore/controller/BookController.java (hunk 12 above)
Comment (Elena-Bruyako) [reply src/main/java/mate/academy/bookstore/controller/BookController.java:20]: not fixed

File: src/main/java/mate/academy/bookstore/service/impl/BookServiceImpl.java [hunk 13]
```diff
@@ -0,0 +1,56 @@
...
+    @Override
+    public BookDto findById(long id) {
+        Book book = bookRepository.findById(id).orElseThrow(
+                () -> new EntityNotFoundException("Can't find book by id: " + id));
+        return bookMapper.toDto(book);
+    }
+
+    @Override
+    public BookDto update(Long id, CreateBookRequestDto createBookRequestDto) {
+        Optional<Book> bookById = bookRepository.findById(id);
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/bookstore/service/impl/BookServiceImpl.java:43]:
```suggestion
        Book book = bookRepository.findById(id).otElseThrow(...;
```

File: src/main/java/mate/academy/bookstore/service/impl/BookServiceImpl.java [hunk 14]
```diff
@@ -0,0 +1,56 @@
...
+        if (bookById.isEmpty()) {
+            throw new EntityNotFoundException("Can't find book by id: " + id);
+        }
+        Book book = bookMapper.toModel(createBookRequestDto);
+        book.setId(id);
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/bookstore/service/impl/BookServiceImpl.java:48]:
```suggestion
        bookMapper.updateBookFromDto(createBookRequestDto, book);
```

**Review by Elena-Bruyako — CHANGES_REQUESTED** [review]
see comments

**Review by Elena-Bruyako — APPROVED** [review]

**Review by MentorMateAcademy — CHANGES_REQUESTED** [review]
you have changes from previous task

**Pull request merged**